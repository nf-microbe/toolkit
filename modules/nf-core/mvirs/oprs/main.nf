process MVIRS_OPRS {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/mvirs:1.1.1--pyhdfd78af_0' :
        'biocontainers/mvirs:1.1.1--pyhdfd78af_0' }"

    input:
    tuple val(meta) , path(fastq)
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(index)

    output:
    tuple val(meta), path("${prefix}.mvirs.bam")        , emit: bam
    tuple val(meta), path("${prefix}.mvirs.fasta.gz")   , emit: fasta
    tuple val(meta), path("${prefix}.mvirs.oprs")       , emit: oprs
    tuple val(meta), path("${prefix}.mvirs.clipped")    , emit: clipped
    tuple val(meta), path("${prefix}.mvirs.prophages")  , emit: prophage
    path "versions.yml"                                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args   ?: ''
    prefix   = task.ext.prefix ?: "${meta.id}"
    """
    mvirs \\
        oprs \\
        -f ${fastq[0]} \\
        -r ${fastq[1]} \\
        -db ${index[0].getBaseName()} \\
        -o ${prefix}.mvirs \\
        -t ${task.cpus} \\
        ${args} || if grep -q "Finishing mVIRs" .command.log; then true; else false; fi

    ### Extract fasta header information into TSV
    grep "^>" ${prefix}.mvirs.fasta \\
        | sed 's/^>//; s/\t/ /; s/:/ /; s/-/ /g; s/OPRs=//; s/HSs=//; s/SF=//' \\
        | awk '{print \$1"\\t"\$2"\\t"\$3"\\t"\$4"\\t"\$5"\\t"\$6}' > ${prefix}.mvirs.prophages  || true

    gzip -f ${prefix}.mvirs.fasta

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mvirs: \$(echo \$(mvirs 2>&1) | sed 's/.*Version: //; s/ .*//')
    END_VERSIONS
    """

    stub:
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.mvirs.bam
    echo "" | gzip > ${prefix}.mvirs.fasta.gz
    touch ${prefix}.mvirs.oprs
    touch ${prefix}.mvirs.clipped
    touch ${prefix}.mvirs.prophages

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mvirs: \$(echo \$(mvirs 2>&1) | sed 's/.*Version: //; s/ .*//')
    END_VERSIONS
    """
}
