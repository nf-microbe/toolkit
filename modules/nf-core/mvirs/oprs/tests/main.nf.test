nextflow_process {

    name "Test Process MVIRS_OPRS"
    script "../main.nf"
    process "MVIRS_OPRS"

    tag "modules"
    tag "modules_nfcore"
    tag "mvirs"
    tag "mvirs/index"
    tag "mvirs/oprs"

    test("fasta.gz") {
        setup {
            run("MVIRS_INDEX") {
                script "../../index/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test' ], // meta map
                        file("${projectDir}/assets/test-datasets/genomics/provirus/illumina/fasta/np_salmoLT2.fasta.gz", checkIfExists: true)
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = [
                    [ id:'test2', single_end:false ],
                    [
                        file("${projectDir}/assets/test-datasets/genomics/provirus/illumina/fastq/ERR4552622_100k_copy_1.fastq.gz", checkIfExists: true),
                        file("${projectDir}/assets/test-datasets/genomics/provirus/illumina/fastq/ERR4552622_100k_copy_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id:'test' ], // meta map
                    file("${projectDir}/assets/test-datasets/genomics/provirus/illumina/fasta/np_salmoLT2.fasta.gz", checkIfExists: true)
                ]
                input[2] = MVIRS_INDEX.out.index
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("fasta.gz - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test2', single_end:false ],
                    [
                        file("${projectDir}/assets/test-datasets/genomics/provirus/illumina/fastq/ERR4552622_100k_copy_1.fastq.gz", checkIfExists: true),
                        file("${projectDir}/assets/test-datasets/genomics/provirus/illumina/fastq/ERR4552622_100k_copy_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id:'test' ], // meta map
                    file("${projectDir}/assets/test-datasets/genomics/provirus/illumina/fasta/np_salmoLT2.fasta.gz", checkIfExists: true)
                ]
                input[2] = [ [:], [] ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
