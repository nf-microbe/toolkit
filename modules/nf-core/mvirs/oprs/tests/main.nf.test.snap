{
    "fasta.gz - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.oprs:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.clipped:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.prophages:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    "versions.yml:md5,bdc8bb39d850e5464d42d6bd7df9a2d2"
                ],
                "bam": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "clipped": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.clipped:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "fasta": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "oprs": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.oprs:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "prophage": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.prophages:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bdc8bb39d850e5464d42d6bd7df9a2d2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-15T10:29:26.478003683"
    },
    "fasta.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.bam:md5,ff71d51c6fdd7c861dfc2291e696ea64"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.fasta.gz:md5,9e1e21890100e9b020a33231a513a2a1"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.oprs:md5,6ee89377971c3156b03db4fdf2b3a0a0"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.clipped:md5,e52b358cb85381ea526452dae127d38d"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.prophages:md5,70ce813cbba0d016917d380ea2cf6af4"
                    ]
                ],
                "5": [
                    "versions.yml:md5,bdc8bb39d850e5464d42d6bd7df9a2d2"
                ],
                "bam": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.bam:md5,ff71d51c6fdd7c861dfc2291e696ea64"
                    ]
                ],
                "clipped": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.clipped:md5,e52b358cb85381ea526452dae127d38d"
                    ]
                ],
                "fasta": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.fasta.gz:md5,9e1e21890100e9b020a33231a513a2a1"
                    ]
                ],
                "oprs": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.oprs:md5,6ee89377971c3156b03db4fdf2b3a0a0"
                    ]
                ],
                "prophage": [
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.mvirs.prophages:md5,70ce813cbba0d016917d380ea2cf6af4"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bdc8bb39d850e5464d42d6bd7df9a2d2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-15T10:30:06.053205854"
    }
}